src/Microsoft.NetCore.Analyzers/Core/Runtime/SpecifyIFormatProvider.cs

```diff
@@ -176,6 +180,13 @@ public override void Initialize(AnalysisContext analysisContext)
                         // report the diagnostic on the overload with CultureInfo as the last parameter, to match the behavior of FxCop.
                         var correctOverload = correctOverloads.FirstOrDefault(overload => overload.Parameters.Last().Type.Equals(iformatProviderType)) ?? correctOverloads.FirstOrDefault();
 
+                        // Use a different mechanism for TryParse methods as they usually have more than 1 extra parameter and
```

Comment:
I don't believe `TryParse` methods should be special cased this way, neither in the code, nor in the option. I think instead we should have 2 modes:

1. Current behavior: Flags when invoked method has an overload with one additional `IFormatProvider` parameter as the last parameter and the remainder of parameters are same.
2. New behavior: Flags when invoked method has an overload with one additional `IFormatProvider` parameter as the last non-ref and non-out parameter and the remainder of parameters (including the following ref and out parameters) are the same.

Reply:
I started my implementation with what you describe in point 2 but I have started to face some issues with some types like numbers, dates... `DateTime.TryParse(string, IFormatProvider, DateTimeStyles, out DateTime)` where there is an extra parameter `XXXStyles`. I was not sure if it would be better to ignore `out` as last and an enum whose name ends in `Styles` as before last or go with my approach. Let me know your preference.

Regarding the option, are you suggesting an enum or to keep a boolean but to have a different name (something like `extended`)?

Reply:
Having put more thought in this, I think the TryParse case above falls into the more generic analysis:

- I have `n` different overloads for a method for similar functionality which were added over a time period.
- I would prefer to have an analyzer that allows me to flag use of non-preferred overloads
- Analyzer would have some default set of BCL APIs where we already know the preferred and non-preferred overloads. In our case, this will be the `TryParse` methods.
- Analyzer would have an option to allow me to specify fully qualified signatures of preferred (and/or non-preferred) overloads for custom APIs.

The above seems to very closely match the [BannedApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md), though that analyzer does not special case banned APIs from BCL libraries and is a much more stricter version of _banned_ as opposed to _non-preferred_ APIs. I am wondering if we should add a new rule to that analyzer with a different ID and much less stricter message, that only deals with preferred and non-preferred symbol pairs, special cases the APIs that we already know about (TryParse group) and allows users to specify an additional customized preferences in `BannedSymbols.txt`. Building on the example from [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md):

```csharp
class MyType
{
   void Method(T1 t);
   void Method(T2 t);  // preferred overload
}
```

Potential "BannedSymbols.txt":
```
MyType.Method(T1);prefer:MyType.Method(T2)
```

@drewnoakes @sharwell for additional comments.

Reply:
@mavasani I think your suggestion is very nice and would definitely be a valuable new rule. Yet I don't think that the `TryParse` method should differ from `Parse` methods, meaning that if we have the other rule then this one shall be made fully obsolete. Methods with a culture should always be favored against non-culture overloads.

Reply:
That is a reasonable argument from an end user experience perspective. However, I don't like us special casing the method in the code and in the option name. What if some other API tomorrow adds a similar overload with some different additional parameter (not styles) and we run into this same problem again? I do believe we need a more generic solution here, even if it is implemented in this analyzer itself.

Reply:
I guess my comment wasn't really clear. My point was to say that I am fully aligned with you and I am just thinking that if/when we have the new rule (assuming this is a different one) then this rule no longer really have value and it would be better to make it obsolete. The new rule with preferred API will most likely favor `Parse` overloads with the culture.

Reply:
@mavasani Shall I create a new issue for the new rule proposal and try to come up with a first implementation?

Reply:
See https://github.com/dotnet/roslyn-analyzers/issues/3188

I still think that it might be nice to have this PR merged and to deprecate this specific analyzer when the related ticket (the new analyzer) is implemented.

WDYT @mavasani ?

Reply:
I don't believe we should have an option targeting just specific set of methods. Can you update the implementation to not have this option at all to start with? There is a risk of this being a breaking change, but it seems a reasonable one.

Reply:
Just to be sure, you want me to remove the following option https://github.com/dotnet/roslyn-analyzers/pull/3074/files#diff-1b5cdc4937a8cdfc79545945ad6c0783R414-R422 and to have the tryparse methods allowed by default?